core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala

```diff
@@ -0,0 +1,216 @@
+        fail(e.toString)
+      }
+    }
+  }
+
+  /**
+   * Tests ZkUtils. The goal is mainly to verify that the behavior of ZkUtils is
+   * correct when isSecure is set to true.
+   */
+  @Test
+  def testZkUtils() {
+    assertTrue(zkUtils.isSecure)
+    for (path <- zkUtils.persistentZkPaths) {
+      zkUtils.makeSurePersistentPathExists(path)
+      if(!path.equals(ZkUtils.ConsumersPath)) {
+        var stat: Stat = new Stat;
+        val aclList = zkUtils.zkConnection.getZookeeper.getACL(path, stat)
+        assertTrue(aclList.size == 2)
+        for (acl: ACL <- aclList.asScala) {
+          assertTrue(isAclSecure(acl))
+        }
+      }
+    }
+    // Test that can create: createEphemeralPathExpectConflict
+    zkUtils.createEphemeralPathExpectConflict("/a", "")
+    verify("/a")
+    // Test that can create: createPersistentPath
+    zkUtils.createPersistentPath("/b")
+    verify("/b")
+    // Test that can create: createSequentialPersistentPath
+    val seqPath = zkUtils.createSequentialPersistentPath("/c", "")
+    verify(seqPath)
+    // Test that can update: updateEphemeralPath
+    zkUtils.updateEphemeralPath("/a", "updated")
+    val valueA: String = zkUtils.zkClient.readData("/a")
+    assertTrue(valueA.equals("updated"))
+    // Test that can update: updatePersistentPath
+    zkUtils.updatePersistentPath("/b", "updated")
+    val valueB: String = zkUtils.zkClient.readData("/b")
+    assertTrue(valueB.equals("updated"))
+
+    info("Leaving testZkUtils")
+  }
+
+  @Test
+  def testZkMigration() {
+    assertTrue(zkUtils.isSecure)
+    val otherZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    try {
+      info("zkConnect string: %s".format(zkConnect))
+      for (path <- otherZkUtils.securePersistentZkPaths) {
+        info("Creating " + path)
+        otherZkUtils.makeSurePersistentPathExists(path)
+        // Create a child for each znode to exercise the recurrent
+        // traversal of the data tree
+        otherZkUtils.createPersistentPath(path + "/fpjwashere", "")
+      }
+      ZkSecurityMigrator.run(Array("--kafka.go=secure", "--zookeeper.connect=%s".format(zkConnect)))
+      info("Done with migration")
+      for (path <- zkUtils.securePersistentZkPaths) {
+        zkUtils.makeSurePersistentPathExists(path)
+        var stat: Stat = new Stat
+        val listParent = zkUtils.zkConnection.getZookeeper.getACL(path, stat)
+        assertTrue(listParent.size == 2)
+        for (acl: ACL <- listParent.asScala) {
+          assertTrue(path, isAclSecure(acl))
+        }
+        val childPath = path + "/fpjwashere"
+        val listChild = zkUtils.zkConnection.getZookeeper.getACL(childPath, stat)
+        assertTrue("Path and size: %s, %d".format(childPath, listChild.size), listChild.size == 2)
+        for (acl: ACL <- listChild.asScala) {
+          assertTrue(childPath, isAclSecure(acl))
+        }
+      }
+    } finally {
+      otherZkUtils.close()
+    }
+  }
+  
+  @Test
+  def testZkAntiMigration() {
```

Comment:
Perhaps we can share most of the code btw testZkMigration and testZkAntiMigration?